plugins/analytics/public/dockerhub.py

```diff
@@ -0,0 +1,188 @@
+        "name": "DockerImageMetadata",
+        "description": "Fetch metadata from docker hub for a docker image",
+    }
+
+    acts_on: list[ObservableType] = [
+        ObservableType.docker_image,
+    ]
+
+    def _get_observable(self, obs_type, value):
+        cls = observable.TYPE_MAPPING[obs_type]
+        obs = cls.find(value=value)
+        if not obs:
+            obs = cls(value=value).save()
+        return obs
+
+    def _get_context(self, metadata):
+        context = {"source": "hub.docker.com"}
+        if metadata:
+            context["Analysis"] = "Image found"
+            context["Image stats"] = {
+                "Pulls": metadata.get("pull_count", "N/A"),
+                "Last updated": metadata.get("last_updated", "N/A"),
+                "Registered": metadata.get("date_registered", "N/A"),
+                "Stars": metadata.get("star_count", "N/A"),
+            }
+            context["Image details"] = metadata.get("tags", "N/A")
+            context["User details"] = metadata.get("user", "Unknown")
+        else:
+            context["Analysis"] = "Image not found"
+        return context
+
+    def _create_digest_context(self, image_obs, tag_name, image_tag):
+        if not image_tag:
+            return {}
+        fullname = f"{image_obs.value}:{tag_name}"
+        arch = image_tag.get("arch", "N/A")
+        os = image_tag.get("os", "N/A")
+        pushed = image_tag.get("last_pushed", "N/A")
+        context = {"image name": fullname, "arch": f"{os}/{arch}", "pushed": pushed}
+        return context
+
+    def _create_digest_observable(self, image_obs, digest, context, link_type):
+        sha_obs = self._get_observable(observable.ObservableType.sha256, value=digest)
+        if context:
+            sha_obs.add_context("hub.docker.com", context)
+            sha_obs.tag({"dockerhub"})
+        image_obs.link_to(sha_obs, link_type, "")
+        return sha_obs
+
+    def _make_relationships(self, image_obs, metadata):
+        username = metadata.get("user", {}).get("username", "")
+        if username:
+            user_obs = self._get_observable(
+                observable.ObservableType.generic, value=username
+            )
+            user_obs.tag({"type:user_account", "dockerhub"})
+            user_obs.link_to(image_obs, "creates", "")
+        for tag_name, image_tags in metadata.get("tags", {}).items():
+            for image_tag in image_tags:
+                context = self._create_digest_context(image_obs, tag_name, image_tag)
+                digest = re.match(
+                    r"sha256:([0-9a-f]{64})$",
+                    image_tag.get("digest", ""),
+                    re.IGNORECASE,
+                )
+                if digest:
+                    self._create_digest_observable(
+                        image_obs, digest.group(1), context, "results_to"
+                    )
+                for layer in image_tag.get("layers", []):
+                    layer_digest = re.match(
```

Comment:
why regex ?

Reply:
Inititally I used `digest.split(":")`. The advantage of regex is that it also validates the digest event if it is supposed to be `sha256:hex_digest`